zerver/lib/message.py

```diff
@@ -344,3 +347,55 @@ def render_markdown(message, content, realm=None, realm_alert_words=None, messag
         message.is_me_message = Message.is_status_message(content, rendered_content)
 
     return rendered_content
+
+def add_missing_messages(user_profile):
+    # type: (UserProfile) -> None
+    def create_user_message_row(user_profile, message):
+        # type: (UserProfile, Message) -> None
+        UserMessage.objects.create(user_profile=user_profile,
+                                   message=message, flags=0)
+
+    all_subs = Subscription.objects.select_related('recipient').filter(user_profile=user_profile)
+    recipients = [sub.recipient for sub in all_subs]
+    all_messages = Message.objects.select_related().filter(recipient__in=recipients,
+                                                           id__gt=user_profile.last_active_message_id).order_by('pub_date')
+    all_non_stream_msgs = [msg for msg in all_messages if not msg.recipient.type == Recipient.STREAM]
+    all_stream_msgs = [msg for msg in all_messages if msg.recipient.type == Recipient.STREAM]
+
+    # For all non stream messages create the UserMessage rows right away.
+    for msg in all_non_stream_msgs:
+        create_user_message_row(user_profile, msg)
+
+    # For Stream messages we need to check messages against data from
+    # RealmAuditLog for visibility to users.
+    stream_messages = defaultdict(list)  # type: DefaultDict[int, List]
+    for msg in all_stream_msgs:
+        stream_messages[msg.recipient.type_id].append(msg)
+    stream_ids = stream_messages.keys()
+    events = ['subscription_created', 'subscription_deactivated', 'subscription_activated']
+    subscription_logs = RealmAuditLog.objects.filter(modified_user=user_profile,
+                                                     modified_stream__id__in=stream_ids,
+                                                     event_type__in=events).order_by('event_time')
+    streams_subscription_logs = defaultdict(list)  # type: DefaultDict[int, List]
+    for log in subscription_logs:
+        streams_subscription_logs[log.modified_stream.id].append(log)
+    for (stream_id, messages) in stream_messages.items():
```

Comment:
hmm, I guess I am running out of names. I used it before.
Will think of better name though.